Tests are failing on Ubuntu 22.04 and 20.04 with python 3.10 and 3.8 correspondingly

Tests are failing on Ubuntu 22.04 and 20.04 with python 3.10 and 3.8 correspondingly.
I do not know if it is expected or not - i just got emails from PPA and decided to highlight this.
If it is expected from this repo point of view - then ok. Nothing can be done on ppa side regarding that.

https://launchpadlibrarian.net/838408047/buildlog_ubuntu-jammy-amd64.solaar_1.1.18+202512182049~ubuntu22.04.1_BUILDING.txt.gz
https://launchpadlibrarian.net/838407884/buildlog_ubuntu-focal-amd64.solaar_1.1.18+202512182049~ubuntu20.04.1_BUILDING.txt.gz
```
=================================== FAILURES ===================================
_ test_notification_flag_str[NotificationFlag.MOUSE_EXTRA_BUTTONS|LINK_QUALITY-link quality\n               mouse extra buttons] _

flag_bits = <NotificationFlag.MOUSE_EXTRA_BUTTONS|LINK_QUALITY: 525312>
expected_names = 'link quality\n               mouse extra buttons'

    @pytest.mark.parametrize(
        "flag_bits, expected_names",
        [
            (None, ""),
            (hidpp10_constants.NotificationFlag(0x0), "none"),
            (hidpp10_constants.NotificationFlag(0x001000), "multi touch"),
            (hidpp10_constants.NotificationFlag(0x080000), "mouse extra buttons"),
            (
                hidpp10_constants.NotificationFlag(0x080400),
                ("link quality\n               mouse extra buttons"),
            ),
        ],
    )
    def test_notification_flag_str(flag_bits, expected_names):
>       flag_names = hidpp10_constants.flags_to_str(
            hidpp10_constants.NotificationFlag(flag_bits) if flag_bits is not None else None, fallback="none"
        )

tests/logitech_receiver/test_hidpp10.py:292: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
logitech_receiver/hidpp10_constants.py:120: in flags_to_str
    flag_names = NotificationFlag.flag_names(flags)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <enum 'NotificationFlag'>
flags = <NotificationFlag.MOUSE_EXTRA_BUTTONS|LINK_QUALITY: 525312>

    @classmethod
    def flag_names(cls, flags) -> List[str]:
        """Extract the names of the flags from the integer."""
>       return flags.name.replace("_", " ").lower().split("|")
E       AttributeError: 'NoneType' object has no attribute 'replace'

logitech_receiver/hidpp10_constants.py:94: AttributeError
=============================== warnings summary ===============================
../../../../../../usr/lib/python3/dist-packages/evdev/eventio_async.py:89
  /usr/lib/python3/dist-packages/evdev/eventio_async.py:89: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead
    def __anext__(self):

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=========================== short test summary info ============================
FAILED tests/logitech_receiver/test_hidpp10.py::test_notification_flag_str[NotificationFlag.MOUSE_EXTRA_BUTTONS|LINK_QUALITY-link quality\n               mouse extra buttons]
============= 1 failed, 630 passed, 30 skipped, 1 warning in 2.24s =============
E: pybuild pybuild:369: test: plugin distutils failed with: exit code=1: cd '/<<PKGBUILDDIR>>/.pybuild/cpython3_3.10/build'; python3.10 -m pytest tests
dh_auto_test: error: pybuild --test --test-pytest -i python{version} -p 3.10 returned exit code 13
make: *** [debian/rules:13: build] Error 25
dpkg-buildpackage: error: debian/rules build subprocess returned exit status 2
--------------------------------------------------------------------------------
Build finished at 2025-12-18T20:56:04Z

Finished
--------
```

--- a/lib/logitech_receiver/hidpp10_constants.py
+++ b/lib/logitech_receiver/hidpp10_constants.py
@@ -91,9 +91,12 @@ class NotificationFlag(IntFlag):
     @classmethod
     def flag_names(cls, flags) -> List[str]:
         """Extract the names of the flags from the integer."""
-        if flags is None or flags.name is None:
+        if flags is None:
             return []
-        return flags.name.replace("_", " ").lower().split("|")
+        if flags.name is not None:
+            return flags.name.replace("_", " ").lower().split("|")
+        # Python < 3.11: .name is None for composite flags, decompose manually
+        return [m.name.replace("_", " ").lower() for m in cls if m.value and m in flags]
 
     NUMPAD_NUMERICAL_KEYS = 0x800000
     F_LOCK_STATUS = 0x400000

--- a/tests/logitech_receiver/test_hidpp10.py
+++ b/tests/logitech_receiver/test_hidpp10.py
@@ -296,6 +296,31 @@ def test_notification_flag_str(flag_bits, expected_names):
     assert flag_names == expected_names
 
 
+@pytest.mark.parametrize(
+    "flags, expected",
+    [
+        (None, []),
+        (hidpp10_constants.NotificationFlag(0x000900), ["software present", "wireless"]),  # composite flag, .name is None on Python < 3.11
+        (hidpp10_constants.NotificationFlag(0x100000), ["battery status"]),
+        (hidpp10_constants.NotificationFlag(0x080000), ["mouse extra buttons"]),
+    ],
+)
+def test_notification_flag_names(flags, expected):
+    result = hidpp10_constants.NotificationFlag.flag_names(flags)
+    assert result == expected
+
+
+def test_notification_flag_names_none_does_not_crash():
+    """Regression test for #3184: flag_names crashes with AttributeError when
+    flags.name is None, which happens with certain receiver firmware versions
+    (e.g. Nano C52F reporting 0x000900)."""
+    flags_with_none_name = hidpp10_constants.NotificationFlag(0x000900)
+    result = hidpp10_constants.NotificationFlag.flag_names(flags_with_none_name)
+    assert "software present" in result
+    assert "wireless" in result
+    assert isinstance(result, list)
+
+
 def test_get_device_features():
     result = _hidpp10.get_device_features(device_standard)
 

--- a/tests/logitech_receiver/test_hidpp10.py
+++ b/tests/logitech_receiver/test_hidpp10.py
@@ -300,7 +300,8 @@ def test_notification_flag_str(flag_bits, expected_names):
     "flags, expected",
     [
         (None, []),
-        (hidpp10_constants.NotificationFlag(0x000900), ["software present", "wireless"]),  # composite flag, .name is None on Python < 3.11
+        # composite flag, .name is None on Python < 3.11
+        (hidpp10_constants.NotificationFlag(0x000900), ["software present", "wireless"]),
         (hidpp10_constants.NotificationFlag(0x100000), ["battery status"]),
         (hidpp10_constants.NotificationFlag(0x080000), ["mouse extra buttons"]),
     ],

